artiq/test/test_scheduler.py

```diff
@@ -144,129 +144,39 @@ def test_pending_priority(self):
-                "key": "status"
-            },
-            {
-                "path": [0],
-                "action": "setitem",
-                "value": "running",
-                "key": "status"
-            },
-            {
-                "path": [2],
-                "action": "setitem",
-                "value": "preparing",
-                "key": "status"
-            },
-            {
-                "path": [2],
-                "action": "setitem",
-                "value": "prepare_done",
-                "key": "status"
-            },
-            {
-                "path": [0],
-                "action": "setitem",
-                "value": "paused",
-                "key": "status"
-            },
-            {
-                "path": [2],
-                "action": "setitem",
-                "value": "running",
-                "key": "status"
-            },
-            {
-                "path": [2],
-                "action": "setitem",
-                "value": "run_done",
-                "key": "status"
-            },
-            {
-                "path": [0],
-                "action": "setitem",
-                "value": "running",
-                "key": "status"
-            },
-            {
-                "path": [2],
-                "action": "setitem",
-                "value": "analyzing",
-                "key": "status"
-            },
-            {
-                "path": [2],
-                "action": "setitem",
-                "value": "deleting",
-                "key": "status"
-            },
-            {
-                "path": [],
-                "action": "delitem",
-                "key": 2
-            },
-        ]
+        expectRID0 = _get_basic_steps(0, expid_bg, low_priority)
+        expectRID1 = _get_basic_steps(1, expid_empty, high_priority,
+                                      due_date=late)
+        expectRID2 = _get_basic_steps(2, expid_empty, middle_priority,
+                                      due_date=early)
+        expectRID0.insert(4,
+            {"action": "setitem", "key": "status", "value": "paused",
+             "path": [0]})
+        expectRID0.insert(5,
+            {"action": "setitem", "key": "status", "value": "running",
+             "path": [0]})
+
+        expect = [expectRID0, expectRID1, expectRID2]
+
+        last_state_RID0 = {"action": "setitem", "key": "status",
+                           "value": "run_done", "path": [0]}
+        last_state_RID1 = {"action": "setitem", "key": "status",
+                           "value": "preparing", "path": [1]}
+
         done = asyncio.Event()
-        expect_idx = 0
+
         def notify(mod):
-            nonlocal expect_idx
-            self.assertEqual(mod, expect[expect_idx])
-            expect_idx += 1
-            if expect_idx >= len(expect):
+            if type(mod["key"]) is int:
+                rid = mod["key"]
+            else:
+                rid = mod["path"][0]
```

Comment:
What is that doing? There should at least be an explanatory comment in the code.

Reply:
It is for finding the RID of the mod

Reply:
I had guessed as much, but why is it done that way?

Reply:
Because I have separated the expect case into 3 for each RID, then monitor which state does it go through.
This way can know what they are up to without a conflict of slight time delay.